proposals/NNNN-swiftpm-additional-plugin-apis.md

```diff
@@ -0,0 +1,450 @@
+    }
+
+    /// Any regular products defined in this package (except plugin products),
+    /// in the same order as they are specified in the package manifest.
+    public let products: [Product]
+
+    /// Any regular targets defined in this package (except plugin targets),
+    /// in the same order as they are specified in the package manifest.
+    public let targets: [Target]
+}
+
+/// Represents a single product defined in a package. Specializations represent
+/// different types of products.
+public class Product {
+    /// The name of the product, as defined in the package manifest. It's unique
+    /// among the products of the package in which it is defined.
+    public let name: String
+    
+    /// The targets that directly comprise the product, in the order in which
+    /// they are declared in the package manifest. The product will contain the
+    /// transitive closure of the these targets and their depdendencies.
+    public let targets: [Target]
+}
+
+/// Represents an executable product defined in a package.
+public class ExecutableProduct: Product {
+    /// The target that contains the main entry point of the executable. Every
+    /// executable product has exactly one main executable target. This target
+    /// will always be one of the targets in the product's `targets` array.
+    public let mainTarget: Target
+}
+
+/// Represents a library product defined in a package.
+public class LibraryProduct: Product {
+    /// Whether the library is static, dynamic, or automatically determined.
+    public let type: LibraryType
+}
+
+/// A type of library product.
+public enum LibraryType {
+    /// Static library.
+    case `static`
+
+    /// Dynamic library.
+    case `dynamic`
+
+    /// The type of library is unspecified and will be determined at build time.
+    case `automatic`
+}
+
+/// Represents a single target defined in a package. Specializations represent
+/// different types of targets.
+public class Target {
+    /// The name of the target, as defined in the package manifest. It's unique
+    /// among the targets of the package.
+    public var name: String
+    
+    /// The absolute path of the target directory in the local file system.
+    public var directory: Path
+    
+    /// Any other targets on which this target depends, in the same order as
+    /// they are specified in the package manifest. Conditional dependencies
+    /// that do not apply have already been filtered out.
+    public var dependencies: [Dependency]
+
+    /// Represents a dependency of a target on a product or on another target.
+    public enum Dependency {
+        /// A dependency on a target in the same package.
+        case target(Target)
+
+        /// A dependency on a product in another package.
+        case product(Product)
+    }
+}
+
+/// Represents a target consisting of a source code module, containing either
+/// Swift or source files in one of the C-based languages.
+public class SourceModuleTarget: Target {
```

Comment:
Do we need to retain additional build settings here, such as linker flags?

Reply:
same comment regarding class inheritance

Reply:
> Do we need to retain additional build settings here, such as linker flags?

You mean the ones declared by the package?  Those makes sense, although I wonder how much we need to provide here vs in future proposals that extend this even further.

Any flags derived by a build system would not really make sense here, but the declared ones, I can see.

Reply:
> same comment regarding class inheritance

Understood.  Same reply about identity vs value semantics applies, and also question about what practical impact this would have on plugin authors.

Reply:
I realize now one concrete reason for making this a class — I wanted to nest `Dependency` under `Target`, and I don't think it's possible to define an enum inside a protocol.  Of course, an alternative would be to have `PackageDependency` and `TargetDependency` as top-level entities.  It was just stylistic to have the type be a subtype as opposed to a top-level type, but it's unfortunate to have to prefix the name.

Reply:
I'm going to split these out into `PackageDependency` and `TargetDependency` as top-level entities.  We already have `LibraryType` and `FileType` and so it might be considered more consistent anyway for all the types to be unnested.

Reply:
> It was just stylistic to have the type be a subtype as opposed to a top-level type, but it's unfortunate to have to prefix the name.

we can have a typealias to help with that

Reply:
> Understood. Same reply about identity vs value semantics applies, and also question about what practical impact this would have on plugin authors.

=> If what we are after is reference semantics than making these classes makes sense. how / where do we use such reference semantics tho?

=> even if we go with reference semantics / classes, I think we should avoid inheritance and define the shared part in a protocol. inheritance more often than not leads to dead-ends and having a protocol to define the shared requirements with composition of internal types for the shared implementation if usually more flexible.